solarforecastarbiter/metrics/deterministic.py

```diff
@@ -209,3 +213,113 @@ def centered_root_mean_square(y_true, y_pred):
     return np.sqrt(np.mean(
         ((y_pred - np.mean(y_pred)) - (y_true - np.mean(y_true))) ** 2
     ))
+
+
+def kolmogorov_smirnov_integral(y_true, y_pred, normed=False):
+    """Kolmogorov-Smirnov Test Integral (KSI).
+
+    Parameters
+    ----------
+    y_true : array-like
+        True values.
+    y_pred : array-like
+        Predicted values.
+    normed : bool, optional
+        If True, return the normalized KSI [%].
+
+    Returns
+    -------
+    ksi : float
+        The KSI between the true and predicted values.
+
+    Notes
+    -----
+    The calculation of the empirical CDF uses a right endpoint rule (the
+    default of the statsmodels ECDF function). For example, if the data is
+    [1.0, 2.0], then ECDF output is 0.5 for any input less than 1.0.
+
+    """
+
+    # empirical CDF
+    ecdf_obs = ECDF(y_true)
+    ecdf_fx = ECDF(y_pred)
+
+    # evaluate CDFs
+    x = np.unique(np.concatenate((y_true, y_pred)))
+    y_o = ecdf_obs(x)
+    y_f = ecdf_fx(x)
+
+    # compute metric
+    D = np.abs(y_o - y_f)
+    ksi = np.trapz(D, x=x)
```

Comment:
`ksi = np.sum(D[:-1] * np.diff(x))` I haven't found a numpy function that does what we need.  This formula only works with right-closed intervals (the current default for `ECDF`) but it may be safer to add `side='right'` to the ECDF call, to be explicit.

The integrand's value between x_i and x_i+1 is D(x_i), so the integral is Sum(_ x_0 ^ x_n-1) D(x_i) * (x_i+1 - x_i). For x_n-1, x_n, the two ECDFs will be equal by construction, so that interval adds 0 to the integral.

Reply:
That produces the correct behavior. I'll read through the original KSI paper to see why they're using trapezoid integration, but I think this change is good for now.